File: NamelessInteractive.Insured.Domain/Organisation.fs
module NamelessInteractive.Insured.Domain.Organisation

open NamelessInteractive.Insured.Domain.Shared
open NamelessInteractive.Insured.Domain.Document
open NamelessInteractive.Insured.Domain.Party
open System.ComponentModel.DataAnnotations

type Industry =
    {
        Id: Identifier
        Code: string
        Description: string
    }

type IndustryClassification =
    {
        Id: Identifier
        Code: string
        Description: string
        Category: string
        IndustryCode: string
    }

type OrganisationType =
    {
        Id: Identifier
        Code: string
        Description: string
    }

type BEELevel =
    {
        Id: Identifier
        Code: string
        Description: string
    }

type BEELevelDocument = 
    {
        Id: Identifier
        Validity: ValidityInformation
        Document: Document
    }

type Director = 
    {
        Id: Identifier
        Validity: ValidityInformation
        Party: Party
    }

type CompanyRegistrationNumber = CompanyRegistrationNumber of CompanyRegistrationNumber:string
type VATNumber = VATNumber of VATNumber :string
type PBOReferenceNumber = PBOReferenceNumber of PBOReferenceNumber:string
type IncomeTaxNumber = IncomeTaxNumber of IncomeTaxNumber:string

type CompanyName = 
    {
        [<Display(Name="Registered Name")>]
        [<Required>]
        RegisteredName: string
        [<Display(Name="Trading Name")>]
        TradingName: string
    }

type Organisation =
    {
        Id: Identifier
        CompanyName: CompanyName
        [<Display(Name="Company Registration Number")>]
        CompanyRegistrationNumber: CompanyRegistrationNumber option
        [<Display(Name="VAT Number")>]
        VATNumber: VATNumber option
        [<Display(Name="PBO Reference Number")>]
        PBOReferenceNumber: PBOReferenceNumber option
        [<Display(Name="Income Tax Number")>]
        IncomeTaxNumber: IncomeTaxNumber option
        [<Display(Name="Email Address")>]
        EmailAddress: EmailAddress option
        [<Display(Name="Website")>]
        Website: WebsiteUrl option
        [<Display(Name="Turnover")>]
        Turnover: decimal option
        [<Display(Name="Foreign")>]
        ForeignIndicator: bool
        [<Display(Name="Listed")>]
        ListedIndicator: bool
        [<Display(Name="Parent Organisation")>]
        ParentOrganisation: Organisation option
        [<Display(Name="Organisation Type")>]
        OrganisationType: OrganisationType
        [<Display(Name="Industry")>]
        Industry: Industry option
        [<Display(Name="Industry Classification")>]
        IndustryClassification: IndustryClassification option
        [<Display(Name="Directors")>]
        Directors: Director seq
        [<Display(Name="BEE Level Documents")>]
        BEELevelDocuments: BEELevelDocument seq
    }
    [<Display(Name="Parent Organisation")>]
    member this.ParentOrganisationName 
        with get() = 
            match this.ParentOrganisation with
            | None -> "None"
            | Some(org) -> org.CompanyName.RegisteredName

File: NamelessInteractive.Insured.Web/Scripts/App/Organisation/ViewModels/Organisation.ts
module InsuredApp {
    export module ViewModels {
        export module Organisation {
            export class Organisation {
                Id: Core.Identifier;
                CompanyName: CompanyName;
                CompanyRegistrationNumber: CompanyRegistrationNumber;
                VATNumber: VATNumber;
                PBOReferenceNumber: PBOReferenceNumber;
                IncomeTaxNumber: IncomeTaxNumber;
                EmailAddress: Shared.EmailAddress;
                Website: Shared.WebsiteUrl;
                Turnover: number;
                ForeignIndicator: boolean;
                ListedIndicator: boolean;
                ParentOrganisation: Organisation;
                ParentOrganisationName: string;
                OrganisationType: OrganisationType;
                Industry: Industry;
                IndustryClassification: IndustryClassification;
                Directors: Array<Director>;
                BEELevelDocuments: Array<BEELevelDocument>;
                constructor() {
                    this.CompanyRegistrationNumber = new CompanyRegistrationNumber(null);
                    this.VATNumber = new VATNumber(null);
                    this.PBOReferenceNumber = new PBOReferenceNumber(null);
                    this.IncomeTaxNumber = new IncomeTaxNumber(null);
                }

                private static IfNull(nullTestValue, value, defaultValue) {
                    return Utilities.Shared.ParseUtilities.IfNull(nullTestValue, value, defaultValue);
                }

                static Parse(data: Organisation) {
                    var result = new Organisation();
                    result.Id = new Core.Identifier(data.Id.Identifier);
                    result.CompanyName = CompanyName.Create(data.CompanyName.RegisteredName, data.CompanyName.TradingName);
                    result.CompanyRegistrationNumber.CompanyRegistrationNumber = Organisation.IfNull(data.CompanyRegistrationNumber, data.CompanyRegistrationNumber.CompanyRegistrationNumber, null);
                    result.VATNumber.VATNumber = Organisation.IfNull(data.VATNumber, data.VATNumber.VATNumber, null);
                    result.PBOReferenceNumber.PBOReferenceNumber = Organisation.IfNull(data.PBOReferenceNumber, data.PBOReferenceNumber.PBOReferenceNumber, null);
                    result.IncomeTaxNumber.IncomeTaxNumber = Organisation.IfNull(data.IncomeTaxNumber, data.IncomeTaxNumber.IncomeTaxNumber, null);
                    result.EmailAddress = new Shared.EmailAddress(Organisation.IfNull(data.EmailAddress, data.EmailAddress.EmailAddress, null));
                    result.Website = new Shared.WebsiteUrl(Organisation.IfNull(data.Website, data.Website.WebsiteUrl, null));
                    result.Turnover = data.Turnover;
                    result.ForeignIndicator = data.ForeignIndicator;
                    result.ListedIndicator = data.ListedIndicator;
                    result.ParentOrganisationName = data.ParentOrganisationName;
                    result.OrganisationType = OrganisationType.Create(data.OrganisationType.Id.Identifier, data.OrganisationType.Code, data.OrganisationType.Description);
                    if (data.Industry != null && data.Industry != undefined) {
                        result.Industry = Industry.Create(data.Industry.Id.Identifier, data.Industry.Code, data.Industry.Description);
                    }
                    if (data.IndustryClassification != null && data.IndustryClassification != undefined) {
                        result.IndustryClassification = IndustryClassification.Create(data.IndustryClassification.Id.Identifier, data.IndustryClassification.Code, data.IndustryClassification.Description, data.IndustryClassification.Category, data.IndustryClassification.IndustryCode);
                    }
                    return result;
                }

                static Test() {
                    var result = new Organisation();
                    result.Id = new Core.Identifier(100);
                    result.CompanyName = new CompanyName();
                    result.CompanyName.TradingName = "Trading Company";
                    result.CompanyName.RegisteredName = "Registered Company";
                    return result;
                }
            }
        }
    }
}